File: redundant-connection/redundant-connection.java
class Solution {
    int result[]=new int[2];
    public int getParent(int parent[],int node)
    {
        while(node!=parent[node])
            node=parent[node];
        return node;
    }
    public boolean find(int i,int j,int parent[])
    {
        return getParent(parent,i)==getParent(parent,i);
    }
    public boolean union(int parent[],int i,int j,int count[])
    {
        int Parent_i=getParent(parent,i);
        int Parent_j=getParent(parent,j);
        if(Parent_i==Parent_j) 
        {
            result[0]=i;
            result[1]=j;
            return true;
        }
        if(count[Parent_i]>count[Parent_j])
        {
            parent[Parent_j]=Parent_i;
            count[Parent_i]+=count[Parent_j];
        }
        else
        {
            parent[Parent_i]=Parent_j;
            count[Parent_j]+=count[Parent_i];
        }
        return false;
    }
    public int[] findRedundantConnection(int[][] edges) {
        if(edges==null) return new int[0];
        int n=edges.length+1;
        int parent[]=new int[n];
        int count[]=new int[n];
        
        for(int i=1;i<n;i++){
            parent[i]=i;
            count[i]=1;
        }
        
        for(int edge[]:edges)
        {
            if(union(parent,edge[0],edge[1],count))
                break;
        }
        return result;
    }
}

File: 684-redundant-connection/redundant-connection.cpp
class UnionFind {
    vector<int> parent, rank;
public:
    UnionFind(int size) {
        parent.resize(size);
        rank.resize(size, 0); 
        for (int i = 0; i < size; i++) {
            parent[i] = i;
        }
    }

    int find(int i) {
        if (parent[i] != i) {
            parent[i] = find(parent[i]);
        }
        return parent[i];
    }

    void unite(int i, int j) {
        int rootI = find(i);
        int rootJ = find(j);

        if (rootI != rootJ) {
            
            if (rank[rootI] > rank[rootJ]) {
                parent[rootJ] = rootI;
            } else if (rank[rootI] < rank[rootJ]) {
                parent[rootI] = rootJ;
            } else {
                parent[rootJ] = rootI;
                rank[rootI]++;
            }
        }
    }
};
class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int n = edges.size();
        UnionFind uf(n+1);
        for(auto i : edges){
            int p1 = uf.find(i[0]);
            int p2 = uf.find(i[1]);
            if(p1==p2){
                return i;
            }else{
                uf.unite(i[0], i[1]);
            }
        }
        
        return {};
    }
};